buildtest/buildsystem/base.py

```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
         lines += [self.recipe.get("run")]
 
         return lines
+
+
+class CompilerBuilder(BuilderBase):
+    type = "compiler"
+
+    # Fortran Extensions Links:
+    # https://software.intel.com/content/www/us/en/develop/documentation/fortran-compiler-developer-guide-and-reference/top/compiler-setup/using-the-command-line/understanding-file-extensions.html
+    # Fortran Extensions: http://fortranwiki.org/fortran/show/File+extensions
+    lang_ext_table = {
+        ".c": "C",
+        ".cc": "C++",
+        ".cxx": "C++",
+        ".cpp": "C++",
+        ".c++": "C++",
+        ".f90": "Fortran",
+        ".F90": "Fortran",
+        ".f95": "Fortran",
+        ".f03": "Fortran",
+        ".f": "Fortran",
+        ".F": "Fortran",
+        ".FOR": "Fortran",
+        ".for": "Fortran",
+        ".FTN": "Fortran",
+        ".ftn": "Fortran",
+    }
+
+    cc = None
+    cxx = None
+    fc = None
+    ldflags = None
+    cflags = None
+    cxxflags = None
+    fflags = None
+    cppflags = None
+
+    def _build_testcontent(self):
+        """This method will build the test content from a Buildspec that uses compiler schema. We need a 'compiler'
+           and 'source' key which specifies the source files to compile. We resolve the source file path which can
+           be an absolute value or relative path with respect to Buildspec. The file extension of sourcefile is used
+           to detect the Programming Language which is used to lookup the compiler wrapper based on Language + Compiler.
+           During compiler detection, we set class variables ``self.cc``, ``self.cxx``. ``self.fc``, ``self.cflags``,
+           ``self.cxxflags``, ``self.fflags``, ``self.cppflags``. ``self.ldflags``. Finally we generate the compile
+           command and add each instruction to ``lines`` which contains content of test. Upon completion, we return
+           a list that contains content of the test.
+        """
+
+        self.compiler_recipe = self.recipe.get("compiler")
```

Comment:
Is the compiler key guaranteed to be defined?

Reply:
Same question for source.

Reply:
both compiler and source are required see
https://github.com/shahzebsiddiqui/buildtest-1/blob/compiler_schema/buildtest/buildsystem/schemas/compiler/compiler-v0.0.1.schema.json#L6

https://github.com/shahzebsiddiqui/buildtest-1/blob/6fecf4c42212ab16aec9b18ba35d73c41cf5f3ac/buildtest/buildsystem/schemas/compiler/compiler-v0.0.1.schema.json#L33

The ``source`` key is required with oneOf the compiler names at the moment. For now this was what we discussed in #296